File: finance-tracker-backend/src/main/java/com/finance/tracker/accounts/domain/entities/Account.java
package com.finance.tracker.accounts.domain.entities;

import com.finance.tracker.accounts.domain.*;
import com.finance.tracker.transactions.domain.Currency;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

@Entity
@Getter
@Setter
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Table(name = "accounts", indexes = {
        @Index(name = "idx_account_user", columnList = "user_id"),
        @Index(name = "idx_account_status", columnList = "status")
})
public class Account {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String accountName;

    @Enumerated(EnumType.STRING)
    private AccountType accountType;

    @Column(length = 4, nullable = false)
    private String lastFour;

    @Enumerated(EnumType.STRING)
    private Currency currency;

    private LocalDate openingDate;

    // Use explicit defaults
    @Builder.Default
    private boolean active = true;

    @Builder.Default
    private boolean readOnly = false;

    @CreationTimestamp
    @Column(updatable = false)
    private Instant createdAt;

    private Instant closedAt;

    private String notes;

    @Enumerated(EnumType.STRING)
    private AccountCategory category;

    // Concurrency control: Crucial for financial balance updates
    @Version
    private Long version;

    // Precision for financial accuracy (19 digits total, 4 after decimal)
    @Column(precision = 19, scale = 2)
    private BigDecimal startingBalance;

    @Column(precision = 19, scale = 2)
    private BigDecimal currentBalance;

    @Column(precision = 19, scale = 2)
    private BigDecimal currentOutstanding;

    @Column(precision = 19, scale = 2)
    private BigDecimal creditLimit;

    private Integer statementDayOfMonth; // Changed from String
    private Integer dueDayOfMonth;       // Changed from String

    private Instant balanceAsOf;

    // Use UUID for foreign keys to ensure DB performance
    @Column(nullable = false)
    private UUID userId;

    @Enumerated(EnumType.STRING)
    private AccountStatus status;

    public boolean isAsset() {
        return category == AccountCategory.ASSET;
    }

    public boolean isLiability() {
        return category == AccountCategory.LIABILITY;
    }
}

File: finance_app/lib/features/account/data/model/account.dart
import 'package:finance_app/features/account/data/model/account_category.dart';

import 'account_type.dart';

class Account {
  final String id;
  final String accountName;
  final AccountType accountType;
  final String? lastFour;
  final String currency;
  final DateTime? openingDate;
  final double? startingBalance;
  final double? currentOutstanding;
  final String? statementDayOfMonth;
  final String? dueDayOfMonth;
  final double? creditLimit;
  final double? currentBalance;
  final bool active;
  final bool readOnly;
  final DateTime createdAt;
  final DateTime? closedAt;
  final String? notes;
  final String category;

  Account({
    required this.id,
    required this.accountName,
    required this.accountType,
    this.lastFour,
    required this.currency,
    this.openingDate,
    this.startingBalance,
    this.currentOutstanding,
    this.statementDayOfMonth,
    this.dueDayOfMonth,
    this.creditLimit,
    this.currentBalance,
    required this.active,
    required this.readOnly,
    required this.createdAt,
    this.closedAt,
    this.notes,
    required this.category,
  });

  factory Account.fromJson(Map<String, dynamic> j) {
    return Account(
      id: j['id'],
      accountName: j['accountName'] ?? 'Unnamed Account',
      accountType:
          (j['accountType'] as String?)?.toAccountType() ?? AccountType.unknown,
      lastFour: j['lastFour'],
      currency: j['currency'] ?? 'INR',

      openingDate: j['openingDate'] != null
          ? DateTime.tryParse(j['openingDate'])
          : null,

      startingBalance: _toDouble(j['startingBalance']),
      currentOutstanding: _toDouble(j['currentOutstanding']),
      statementDayOfMonth: j['statementDayOfMonth'],
      dueDayOfMonth: j['dueDayOfMonth'],
      creditLimit: _toDouble(j['creditLimit']),
      currentBalance: _toDouble(j['currentBalance']),

      active: j['active'] ?? true,
      readOnly: j['readOnly'] ?? false,

      createdAt: DateTime.parse(j['createdAt']),
      closedAt: j['closedAt'] != null ? DateTime.tryParse(j['closedAt']) : null,

      notes: j['notes'],

      category: (j['category']),
    );
  }

  double get effectiveBalance {
    if (category == AccountCategory.liability.name.toUpperCase()) {
      return currentOutstanding ?? 0.0;
    }
    if (category == AccountCategory.asset.name.toUpperCase()) {
      return currentBalance ?? startingBalance ?? 0.0;
    }
    return 0.0;
  }

  double get remainingBalance {
    if (category == AccountCategory.liability.name.toUpperCase()) {
      return (creditLimit ?? 0.0) - (currentOutstanding ?? 0.0);
    }
    if (category == AccountCategory.asset.name.toUpperCase()) {
      return currentBalance ?? startingBalance ?? 0.0;
    }
    return 0.0;
  }

  bool isAsset() {
    return category == AccountCategory.asset.name.toUpperCase();
  }
}

double? _toDouble(dynamic v) {
  if (v == null) return null;
  if (v is double) return v;
  if (v is int) return v.toDouble();
  return double.tryParse(v.toString());
}
